File: sorting-searching/sorting-algorithms/Sample-sort/sample-sort.js

function sampleSort(arr, bucketCount) {
  const n = arr.length;

  const samples = [];

  for (let i = 1; i < bucketCount; i++) {
    samples.push(
      arr[Math.floor(i * n / bucketCount)]
    );
  }

  samples.sort((a, b) => a - b);

  const buckets =
    Array.from(
      { length: bucketCount },
      () => []
    );

  for (const num of arr) {
    let bucketIndex = 0;

    while (
      bucketIndex < samples.length &&
      num > samples[bucketIndex]
    ) {
      bucketIndex++;
    }

    buckets[bucketIndex].push(num);
  }

  for (const bucket of buckets) {
    bucket.sort((a, b) => a - b);
  }

  return buckets.flat();
}

const arr = [
  29, 14, 35, 8,
  42, 17, 23, 5, 31
];

console.log("Original array:");
console.log(arr);

const sorted =
  sampleSort(arr, 3);

console.log("\nSorted array:");
console.log(sorted);

File: sorting-searching/sorting-algorithms/Sample-sort/SampleSort.java
import java.util.*;

public class SampleSort {

    static List<Integer> sampleSort(
        int[] arr,
        int bucketCount
    ) {
        int n = arr.length;

        List<Integer> samples =
            new ArrayList<>();

        for (int i = 1; i < bucketCount; i++) {
            samples.add(
                arr[(i * n) / bucketCount]
            );
        }

        Collections.sort(samples);

        List<List<Integer>> buckets =
            new ArrayList<>();

        for (int i = 0; i < bucketCount; i++) {
            buckets.add(new ArrayList<>());
        }

        for (int num : arr) {
            int bucketIndex = 0;

            while (
                bucketIndex < samples.size()
                && num > samples.get(bucketIndex)
            ) {
                bucketIndex++;
            }

            buckets.get(bucketIndex).add(num);
        }

        for (List<Integer> bucket : buckets) {
            Collections.sort(bucket);
        }

        List<Integer> result =
            new ArrayList<>();

        for (List<Integer> bucket : buckets) {
            result.addAll(bucket);
        }

        return result;
    }

    public static void main(String[] args) {
        int[] arr = {
            29, 14, 35, 8,
            42, 17, 23, 5, 31
        };

        System.out.println(
            "Original array:"
        );

        System.out.println(
            Arrays.toString(arr)
        );

        List<Integer> sorted =
            sampleSort(arr, 3);

        System.out.println(
            "\nSorted array:"
        );

        System.out.println(sorted);
    }
}
```
